### PR TITLE
fix(ci): require pre-merge CHANGELOG PR-reference for governed changes

### PR DESCRIPTION
**Pull request by qnbs**
## Problem

`scripts/check-doc-metrics.mjs`'s CHANGELOG completeness gate only enforces a
PR-number reference in `[Unreleased]` **after** squash-merge, once the commit
is on `main` and its subject already carries `(#N)`. Pre-merge, a branch's
own not-yet-squashed commits are (correctly) exempted from that check, since
individual intermediate commits get squashed into one at merge time.

This has left a recurring blind spot: nothing stops a governed PR from
merging without ever adding a CHANGELOG entry, even though the PR's real
number is already knowable via the GitHub API before merge. It has recurred
three times, each requiring a same-pattern follow-up PR to add the missing
reference after the fact:

- #678 → #679
- #684 → #685
- #699 → #700

## Fix

A new, independent pre-merge admission gate:

- `.github/workflows/pr-changelog-reference.yml` — triggers on
  `opened`/`edited`/`synchronize`/`reopened` (title-only edits included).
  Mirrors `pr-text-attribution.yml`'s base-ref self-grading pattern: the
  checker script is loaded from the PR's **base ref**, with a documented
  one-time bootstrap fallback, so a PR cannot weaken the check that grades
  it.
- `scripts/check-pr-changelog-reference.mjs` — fails a governed
  (`feat|fix|perf`) PR unless `CHANGELOG.md`'s `[Unreleased]` section already
  references it as `PR #<N>`, using the PR number from GitHub's own event
  payload (never inferred from git history). Deliberately stricter grammar
  than the existing post-merge bare `#NNN` matcher, since pre-merge there is
  no squash-appended `(#NNN)` to anchor on yet. Fully self-contained (no
  local imports) so the base-ref copy can never break on a missing
  transitive dependency.

The existing `scanUnreleasedTruth` machinery in `check-doc-metrics.mjs` —
governing local pre-push behavior and the historical post-merge/branch-local
exemption — is untouched.

## Non-goals

- Not implementing #675's broader deterministic-identifier-contract scope
  (replacing the unnumbered-commit slug-match fallback with SHA/change-ID
  identifiers). This gate only closes the narrower pre-merge admission gap
  for PRs that already have a real, known PR number — the common case — and
  is fully compatible with #675's eventual direction.
- Not touching local pre-push behavior, since no PR number exists yet
  outside GitHub's own PR-event payload.

## Testing

`tests/unit/checkPrChangelogReference.test.ts` — 17 tests: 13 regression
tests (governed-type grammar incl. scope/`!`, missing/wrong-PR/bare-`#N`/
trailing-`(#N)` rejections, non-governed skip, number-boundary safety,
case/whitespace tolerance) plus 4 fixtures reproducing all three historical
incidents using the real CHANGELOG text from each recovery PR.

Mutation-tested: temporarily weakened the exact-grammar regex to accept bare
`#NNN`, confirmed exactly the two grammar-specific tests failed, restored.

Also verified via the real CLI path (not just the pure function) against
live `GITHUB_EVENT_PATH` fixtures: correct fail-closed, not-a-PR-event
no-op, and no-event-path no-op behavior.

## Summary by Sourcery

Require governed pull requests to include a self-referencing entry in the unreleased changelog before they can merge.

New Features:
- Add a pre-merge CI gate requiring governed feat, fix, and perf pull requests to reference their own PR number in a CHANGELOG.md [Unreleased] bullet.
- Load the admission checker from the PR base ref and use GitHub event metadata to prevent self-weakening and inaccurate PR-number checks.

Bug Fixes:
- Prevent governed pull requests from merging without a corresponding unreleased CHANGELOG entry.

Enhancements:
- Enforce precise PR #N references while handling section boundaries, comments, wrapped bullets, malformed metadata, and non-governed titles safely.

CI:
- Run the CHANGELOG reference check when pull requests are opened, edited, synchronized, or reopened against main.

Documentation:
- Update the CHANGELOG with the new pre-merge admission guard and refresh documented test metrics.

Tests:
- Add regression coverage for valid and invalid references, metadata validation, formatting edge cases, and the three historical missing-reference incidents.

---
## Summary by cubic
Adds a pre-merge CI gate requiring governed (`feat|fix|perf`) PRs to reference their own PR number as `PR #<N>` in `CHANGELOG.md`'s `[Unreleased]` section before merging. Previously the CHANGELOG completeness check only ran after squash-merge, so governed PRs could merge with no entry — a gap that has already required three follow-up PRs (#678→#679, #684→#685, #699→#700).

**Details**
- Runs from the PR's base ref with a bootstrap fallback, so a PR can't weaken the check that grades it; the PR number comes from GitHub's event payload, never git history.
- Requires the stricter `PR #<N>` grammar with a full trailing word boundary, rejecting near-misses like `PR #700alpha` or `PR #700_internal`.
- Only a reference inside an actual bullet entry counts; a PR number in prose, a sub-heading, or any non-indented continuation line won't satisfy the gate.
- Strips HTML comments from the whole changelog before locating the `[Unreleased]` heading, so a commented-out fake section can't satisfy the gate.
- Fails closed on malformed event metadata (non-integer, zero, or negative number; missing or blank title).
- Skips non-governed titles (`docs:`, etc.); local pre-push behavior and the existing post-merge check are untouched.
- Adds regression coverage including real-text fixtures from all three historical incidents, with the test file refactored to reduce duplication; doesn't implement #675's broader identifier contract but stays compatible with it.

<sup>Written for commit b02958dddf6c51e3a5a4de745666e95bff31824f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/qnbs/WorldScript-Studio/pull/705?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **New Features**
  - Added an automated pre-merge check for governed pull requests.
  - Governed pull requests must reference their own number in the `[Unreleased]` changelog section before merging.
  - Pull requests outside the governed categories are not subject to this check.
  - Improved validation to recognize valid changelog entries and reject ambiguous references.

- **Documentation**
  - Documented the new changelog requirement.
  - Updated project test metrics to reflect 7,660+ tests across 605 files.

___

## **CodeAnt-AI Description**
Require governed pull requests to reference themselves in the unreleased changelog before merging

### What Changed
- Pull requests titled as `feat`, `fix`, or `perf` must include a bullet in `CHANGELOG.md`'s `[Unreleased]` section that references their own GitHub PR number.
- Checks run when a PR is opened, edited, updated, or reopened, using the PR number supplied by GitHub.
- Documentation-only and other non-governed PRs continue without this requirement.
- Added coverage for valid references, malformed or unrelated numbers, wrapped changelog bullets, comments, section boundaries, and previous incident cases.

### Impact
`✅ Fewer governed changes merged without changelog entries`
`✅ Earlier CI feedback before merge`
`✅ Accurate self-references tied to the actual PR number`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: scripts/check-pr-changelog-reference.mjs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  if (!heading) return '';
+  const afterHeading = changelog.slice(heading.index + heading[0].length);
+  const nextHeading = afterHeading.search(/^##\s/m);
+  const section = nextHeading === -1 ? afterHeading : afterHeading.slice(0, nextHeading);
+  return section.replace(/<!--[\s\S]*?(?:-->|$)/g, '');
+}
+
+// QNBS-v3: exact "PR #NNN" grammar, stricter than check-doc-metrics.mjs's post-merge bare "#NNN" matcher — pre-merge there is no squash-appended "(#NNN)" to anchor on, so a bare "#NNN" could belong to an unrelated issue/PR mention instead of a genuine self-reference.
+export function isReferencedByPrLabel(prNumber, unreleasedSection) {
+  return new RegExp(`\\bPR\\s*#${prNumber}(?!\\d)`, 'i').test(unreleasedSection);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs]: Fixed in 3ce40e9f — the [Unreleased] check is now scoped to parsed bullet entries (extractBulletEntries), not the raw section text, so a PR number in unrelated prose no longer satisfies the gate.

File: scripts/check-pr-changelog-reference.mjs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    console.error(
+      `[check-pr-changelog-reference] FAIL — CHANGELOG.md's [Unreleased] section does not yet reference "PR #${pr.number}". Add (or update) a bullet describing this change and reference it literally as "PR #${pr.number}" before merging.`,
+    );
+    process.exit(1);
+  }
+  console.log(`[check-pr-changelog-reference] OK — [Unreleased] references PR #${pr.number}`);
+}
+
+// QNBS-v3: only run the CLI side-effect when invoked directly — checkPrChangelogReference stays importable from a unit test.
+if (process.argv[1] === fileURLToPath(import.meta.url)) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs:145]: Verified false: Node.js always resolves process.argv[1] to an absolute path regardless of invocation style, so it equals fileURLToPath(import.meta.url) even for a relative-path invocation. Empirically re-confirmed just now: `node scripts/check-pr-changelog-reference.mjs` (relative path, from the repo root) prints the expected log line and exits 0, proving main() runs. The bootstrap-fallback path in the workflow (which also invokes the relative path) behaves identically.

File: scripts/check-pr-changelog-reference.mjs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+/** Pure decision function — kept separate from I/O so it is directly unit-testable. */
+export function checkPrChangelogReference({ prNumber, prTitle, changelog }) {
+  if (!GOVERNED_COMMIT_TYPE.test(prTitle ?? '')) {
+    return { ok: true, reason: 'not-governed' };
+  }
+  const unreleasedSection = getUnreleasedSectionText(changelog ?? '');
+  return isReferencedByPrLabel(prNumber, unreleasedSection)
+    ? { ok: true, reason: 'referenced' }
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs:85]: Fixed in 3ce40e9f — same fix as the codeant-ai finding above: the check is now scoped to parsed bullet entries, not the whole section.

File: .github/workflows/pr-changelog-reference.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          GITHUB_EVENT_PATH: ${{ github.event_path }}
+        run: |
+          set -euo pipefail
+          BASE="${{ github.event.pull_request.base.sha }}"
+          mkdir -p /tmp/base-scripts
+          CHECKER=scripts/check-pr-changelog-reference.mjs
+          if git show "$BASE:scripts/check-pr-changelog-reference.mjs" > /tmp/base-scripts/check-pr-changelog-reference.mjs 2>/dev/null; then
+            CHECKER=/tmp/base-scripts/check-pr-changelog-reference.mjs
+          else
+            echo "::notice::check-pr-changelog-reference.mjs not found on base ref (bootstrap PR) — using this PR's own copy this one time."
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/pr-changelog-reference.yml:29]: Already addressed in 57878255 — this PR's own [Unreleased] entry references "PR #705" before merge.

File: .github/workflows/pr-changelog-reference.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+          fi
+          node "$CHECKER"
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/pr-changelog-reference.yml:31]: Already addressed in 57878255 — see the reply on the equivalent chatgpt-codex-connector finding.

File: .github/workflows/pr-changelog-reference.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      - name: Require [Unreleased] to reference this PR before merge
+        env:
+          GITHUB_EVENT_PATH: ${{ github.event_path }}
+        run: |
+          set -euo pipefail
+          BASE="${{ github.event.pull_request.base.sha }}"
+          mkdir -p /tmp/base-scripts
+          CHECKER=scripts/check-pr-changelog-reference.mjs
+          if git show "$BASE:scripts/check-pr-changelog-reference.mjs" > /tmp/base-scripts/check-pr-changelog-reference.mjs 2>/dev/null; then
+            CHECKER=/tmp/base-scripts/check-pr-changelog-reference.mjs
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/pr-changelog-reference.yml:27]: Valid observation, but out of scope for this PR: copying only the checker script (not the workflow YAML) from the base ref is the exact same self-grading pattern already established by .github/workflows/pr-text-attribution.yml, which has the identical limitation today (a PR could still edit that .yml itself while keeping the same check name). Hardening workflow-file trust repo-wide (e.g. a CODEOWNERS rule or ruleset restricting .github/workflows/ edits) is an architectural question applying equally to the pre-existing pattern, not something unique to this PR. Tracking as a follow-up rather than expanding this bounded fix to solve a different, broader problem.

File: scripts/check-pr-changelog-reference.mjs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    if (current.length > 0) entries.push(current.join(' '));
+    current = [];
+  };
+  for (const rawLine of unreleasedSection.split('\n')) {
+    const line = rawLine.trim();
+    if (/^-\s/.test(line)) {
+      flush();
+      current.push(line);
+    } else if (current.length > 0 && line !== '') {
+      current.push(line);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs]: Fixed in b0528251 (generalized further in 7cc65d9b to flush on any non-indented line, not just headings).

File: scripts/check-pr-changelog-reference.mjs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      flush();
+    }
+  }
+  flush();
+  return entries;
+}
+
+// QNBS-v3: exact "PR #NNN" grammar, stricter than check-doc-metrics.mjs's post-merge bare "#NNN" matcher — pre-merge there is no squash-appended "(#NNN)" to anchor on, so a bare "#NNN" could belong to an unrelated issue/PR mention instead of a genuine self-reference.
+export function isReferencedByPrLabel(prNumber, text) {
+  return new RegExp(`\\bPR\\s*#${prNumber}(?!\\d)`, 'i').test(text);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs]: Fixed in b0528251 — isReferencedByPrLabel now uses a full word boundary (?!\w) instead of (?!\d), so "PR #705alpha"/"PR #705_" no longer match.

File: scripts/check-pr-changelog-reference.mjs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import { readFileSync } from 'node:fs';
+import process from 'node:process';
+import { fileURLToPath } from 'node:url';
+
+// QNBS-v3: duplicated from check-doc-metrics.mjs's GOVERNED_COMMIT_TYPE (kept in sync manually, not via import) so this file has zero local dependencies — see file header.
+const GOVERNED_COMMIT_TYPE = /^(?:feat|fix|perf)(?:\([^)]*\))?!?:\s*/i;
+
+// QNBS-v3: duplicated from check-doc-metrics.mjs's getUnreleasedSectionText for the same self-containment reason.
+function getUnreleasedSectionText(changelog) {
+  const heading = /^## \[Unreleased\]\s*$/m.exec(changelog);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs]: Fixed in e0a4ef2e — comments are now stripped from the whole document before the [Unreleased] heading search runs, so a commented-out fake heading earlier in the file can no longer hijack the section boundary.

File: scripts/check-pr-changelog-reference.mjs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  };
+  for (const rawLine of unreleasedSection.split('\n')) {
+    const line = rawLine.trim();
+    if (/^-\s/.test(line)) {
+      flush();
+      current.push(line);
+    } else if (/^#{1,6}\s/.test(line)) {
+      flush();
+    } else if (current.length > 0 && line !== '') {
+      current.push(line);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs]: Same fix as the earlier heading-flush finding — see b0528251, generalized further in 7cc65d9b.

File: scripts/check-pr-changelog-reference.mjs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+import { readFileSync } from 'node:fs';
+import process from 'node:process';
+import { fileURLToPath } from 'node:url';
+
+// QNBS-v3: duplicated from check-doc-metrics.mjs's GOVERNED_COMMIT_TYPE (kept in sync manually, not via import) so this file has zero local dependencies — see file header.
+const GOVERNED_COMMIT_TYPE = /^(?:feat|fix|perf)(?:\([^)]*\))?!?:\s*/i;
+
+// QNBS-v3: strips comments from the WHOLE document before searching for the heading — a commented-out template containing a literal "## [Unreleased]" line earlier in the file would otherwise hijack the section boundary, since slicing off the opening "<!--" before comment-removal runs left the fake section's own content unstrippable.
+function getUnreleasedSectionText(changelog) {
+  const withoutComments = changelog.replace(/<!--[\s\S]*?(?:-->|$)/g, '');
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs:24]: Valid theoretical edge case, but this exact comment-stripping regex (/|$)/g) is already used unmodified in scripts/check-doc-metrics.mjs's own getUnreleasedSectionText, which this file deliberately duplicates. A correct fix needs real Markdown code-span/fence awareness — larger scope than this bounded gate, and not something this PR introduces. Tracking as a documented, accepted limitation shared with the pre-existing historical checker.

File: scripts/check-pr-changelog-reference.mjs [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+function extractBulletEntries(unreleasedSection) {
+  const entries = [];
+  let current = [];
+  const flush = () => {
+    if (current.length > 0) entries.push(current.join(' '));
+    current = [];
+  };
+  for (const rawLine of unreleasedSection.split('\n')) {
+    const trimmed = rawLine.trim();
+    if (/^-\s/.test(trimmed)) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs:47]: Declining for now: this requires the changelog author to deliberately embed a decoy bullet inside a fenced code example matching their own PR's exact number — a self-defeating scenario, since an author who wanted to bypass the gate could just write a real bullet instead. A correct general fix needs full CommonMark-aware block-structure parsing (fence-state tracking), out of proportion for this bounded pre-merge check. The already-fixed heading/blockquote cases covered plausible accidental formatting; this requires deliberate adversarial authorship by the same person the gate protects against forgetting the entry. Tracking as an accepted limitation — a follow-up issue for full CommonMark-based parsing would be the right place to close this class completely.

File: scripts/check-pr-changelog-reference.mjs [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-changelog-reference.mjs:47]: Same reasoning as the fenced-code-block finding above — declining for the same reasons (deliberate adversarial authorship required, full parsing out of proportion for this bounded gate).

File: .github/workflows/pr-changelog-reference.yml [hunk 14]
```diff
@@ -0,0 +1,31 @@
+name: PR CHANGELOG Reference Guard
+on:
+  pull_request:
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/pr-changelog-reference.yml:3]: Fixed via repository configuration: added "Require this PR's own number in CHANGELOG.md [Unreleased] before merge" to main's required status checks (alongside the existing ✅ CI Success), so a failure of this workflow now mechanically blocks merge — not just a visible-but-advisory status.

File: .github/workflows/pr-changelog-reference.yml [hunk 15]
```diff
@@ -0,0 +1,31 @@
+name: PR CHANGELOG Reference Guard
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/pr-changelog-reference.yml:1]: Same fix as the chatgpt-codex-connector finding above: this check is now a required status check on main's branch protection, alongside ✅ CI Success.
